symmath/symmath_check.py

```diff
@@ -0,0 +1,337 @@
+
+    ans is student-typed answer. It is expected to be ascii math, but the code
+     below would support a sympy string.
+
+    dynamath is the PMathML string converted by MathJax. It is used if
+     evaluation with ans is not sufficient.
+
+    options is a string with these possible substrings, set as an xml property
+     of the problem:
+     -matrix - make a sympy matrix, rather than a list of lists, if possible
+     -qubit - passed to my_sympify
+     -imaginary - used in formla, presumably to signal to use i as sqrt(-1)?
+     -numerical - force numerical comparison.
+    """
+
+    msg = ''
+    # msg += '<p/>abname=%s' % abname
+    # msg += '<p/>adict=%s' % (repr(adict).replace('<','&lt;'))
+
+    threshold = 1.0e-3   # for numerical comparison (also with matrices)
+    DEBUG = debug
+
+    if xml is not None:
+        DEBUG = xml.get('debug', False)  	# override debug flag using attribute in symbolicmath xml
+        if DEBUG in ['0', 'False']:
+            DEBUG = False
+
+    # options
+    if options is None:
+        options = ''
+    do_matrix = 'matrix' in options
+    do_qubit = 'qubit' in options
+    do_numerical = 'numerical' in options
+
+    # parse expected answer
+    try:
+        fexpect = my_sympify(str(expect), matrix=do_matrix, do_qubit=do_qubit)
+    except Exception as err:  # lint-amnesty, pylint: disable=broad-except
+        msg += Markup('<p>Error {err} in parsing OUR expected answer "{expect}"</p>').format(err=err, expect=expect)
+        return {'ok': False, 'msg': make_error_message(msg)}
+
+    ###### Sympy input #######
+    # if expected answer is a number, try parsing provided answer as a number also
+    try:
+        fans = my_sympify(str(ans), matrix=do_matrix, do_qubit=do_qubit)
+    except Exception as err:  # lint-amnesty, pylint: disable=broad-except, unused-variable
+        fans = None
+
+    # do a numerical comparison if both expected and answer are numbers
+    if hasattr(fexpect, 'is_number') and fexpect.is_number \
+       and hasattr(fans, 'is_number') and fans.is_number:
+        if is_within_tolerance(fexpect, fans, threshold):
+            return {'ok': True, 'msg': msg}
+        else:
+            msg += Markup('<p>You entered: {fans}</p>').format(fans=to_latex(fans))
+            return {'ok': False, 'msg': msg}
+
+    if do_numerical:		# numerical answer expected - force numerical comparison
+        if is_within_tolerance(fexpect, fans, threshold):
+            return {'ok': True, 'msg': msg}
+        else:
+            msg += Markup('<p>You entered: {fans} (note that a numerical answer is expected)</p>').\
+                format(fans=to_latex(fans))
+            return {'ok': False, 'msg': msg}
+
+    if fexpect == fans:
+        msg += Markup('<p>You entered: {fans}</p>').format(fans=to_latex(fans))
+        return {'ok': True, 'msg': msg}
+
+    ###### PMathML input ######
+    # convert mathml answer to formula
```

Comment:
If I recall correctly, we found that everything below this line is poorly test-covered, probably broken, and potentially unused.

That's a great insight we reached that we don't want to lose! Could you leave a comment here? You could even link out to #39 .

Reply:
I think I understand what you want now. I've added the comment on line 258.